### PR TITLE
State cache and other performance optimizations

### LINKED ISSUE #257: Execution optimisation
Execution stack is currently suboptimal. Each call involves querying code from storage, getting runtime version, initialising wasm module, etc. A bulk of work should only be done once per code change. There should be a module between state and executor that tracks code changes and caches reusable bits for each state. If the state does not have code changed since the last state it should reuse cached data for the new state. 

Related: #238

### PR DESCRIPTION
**Pull request by arkpar**
Closes #257 

Major change introduced in this PR is canonical state cache. This works in the same way as in `parity-ethereum`. The cache caches a single state `C` that is considered "canonical" or "best". Any state `S` that is derived from `C` can use cashed value cache as long as the key is not overwritten in any of the blocks between `C` and `S`.

A few other minor changes:

* `changes` in`CallResult` and storage root calculation are not really required for each call. So the whole struct is dropped.
* Executor now queries the code hash from storage/cache instead of querying the code and hashing it every call.
* wasmi module instance is reused for each call. The instance is now cached per thread.

### CODE REVIEW

File: core/executor/src/wasm_executor.rs [hunk 1]
```diff
@@ -638,74 +636,109 @@ impl WasmExecutor {
...
+			&ImportsBuilder::new()
+			.with_resolver("env", FunctionExecutor::<E>::resolver())
+			)?;
+
+		// extract a reference to a linear memory, optional reference to a table
+		// and then initialize FunctionExecutor.
+		let memory = intermediate_instance
+			.not_started_instance()
+			.export_by_name("memory")
+			// TODO: with code coming from the blockchain it isn't strictly been compiled with rustc anymore.
```
Comment (gnunicorn) [inline core/executor/src/wasm_executor.rs]: No TODOs without links to issues in PRs to master: please put the info into a ticket and provide the link here or remove the TODO :) .

Comment (arkpar) [reply core/executor/src/wasm_executor.rs]: This TODO is copy-pasted from a similar call above. I'm not sure what it is about to be honest.

File: core/client/db/src/storage_cache.rs [hunk 2]
```diff
@@ -0,0 +1,404 @@
...
+		}
+	}
+
+	/// Check if the key can be returned from cache by matching current block parent hash against canonical
+	/// state and filtering out entries modified in later blocks.
+	fn is_allowed(key: &[u8], parent_hash: &Option<B::Hash>, modifications: &VecDeque<BlockChanges<B::Header>>) -> bool {
+		let mut parent = match *parent_hash {
+			None => {
+				trace!("Cache lookup skipped for {:?}: no parent hash", key);
+				return false;
```
[the remark this answers is not in the document]
Comment (arkpar) [reply core/client/db/src/storage_cache.rs]: This check would need to be done at each call site anyway. Less repeated code. This is a private function after all.

File: core/client/db/src/storage_cache.rs [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+		enacted: &[B::Hash],
+		retracted: &[B::Hash],
+		changes: Vec<(StorageKey, Option<StorageValue>)>,
+		commit_hash: Option<B::Hash>,
+		commit_number: Option<<B::Header as Header>::Number>,
+		is_best: F,
+	) {
+		let mut cache = self.shared_cache.lock();
+		let is_best = is_best();
+		trace!("Syncing cache, id = (#{:?}, {:?}), parent={:?}, best={}", commit_number, commit_hash, self.parent_hash, is_best);
```
[the remark this answers is not in the document]
Comment (arkpar) [reply core/client/db/src/storage_cache.rs]: Module path is used as target by default. I.e. `substrate_client_db::storage_cache=trace` can be used here to filter all messages specific to this module. IMO explicit target is useful when you need to have a single trace across multiple modules or to shorten the filter name for the most common diagnostics.

File: core/client/db/src/lib.rs [hunk 4]
```diff
@@ -550,6 +561,7 @@ impl<Block: BlockT> Backend<Block> {
 			changes_tries_storage,
 			blockchain,
 			canonicalization_delay,
+			shared_cache: new_shared_cache(STATE_CACHE_SIZE_BYTES),
```
[the remark this answers is not in the document]
Comment (arkpar) [reply core/client/db/src/lib.rs]: Added an issue for now: #1360

File: core/client/db/src/storage_cache.rs [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+	/// Information on the modifications in recently committed blocks; specifically which keys
+	/// changed in which block. Ordered by block number.
+	modifications: VecDeque<BlockChanges<B::Header>>,
+}
+
+pub type SharedCache<B, H> = Arc<Mutex<Cache<B, H>>>;
+
+/// Create new shared cache instance with given max memory usage.
+pub fn new_shared_cache<B: Block, H: Hasher>(shared_cache_size: usize) -> SharedCache<B, H> {
+	let cache_items = shared_cache_size / 100; // Estimated average item size. TODO: more accurate tracking
```
[the remark this answers is not in the document]
Comment (arkpar) [reply core/client/db/src/storage_cache.rs]: Added an issue for now: #1360

File: core/executor/src/wasm_executor.rs (hunk 1 above)
Comment (pepyakin) [reply core/executor/src/wasm_executor.rs]: AFAIR, initially we had an `expect("this code is from rustc ; it's always correct")`, but then somebody actually noted that this code isn't actually coming from a rustc generated wasm artifact but from the blockchain storage and so it became harder to claim that rustc generated this binary.

I'm not sure if we want to panic here, it would be better if we could return an `Err` here.

File: core/executor/src/wasm_executor.rs [hunk 6]
```diff
@@ -638,74 +636,109 @@ impl WasmExecutor {
...
+		let table: Option<TableRef> = module_instance
 			.export_by_name("__indirect_function_table")
 			.and_then(|e| e.as_table().cloned());
 
-		let mut fec = FunctionExecutor::new(memory.clone(), heap_pages, table, ext)?;
-
-		// finish instantiation by running 'start' function (if any).
-		let instance = intermediate_instance.run_start(&mut fec)?;
+		let low = memory.lowest_used();
+		let used_mem = memory.used_size();
```
Comment (pepyakin) [inline core/executor/src/wasm_executor.rs]: Assuming that `used_size` returns the highest written address so far, this will be set to the address of last written byte (shouldn't it be +1?) and then in `Heap` we assume that we can allocate memory starting from this address.

But because we didn't yet run the start function (which could run arbitrary initialization code) this assumption can be violated.

I think this would work for now since rust doesn't use start function at all atm and we only have rust runtimes now. But I'm not sure that this is universal approach, since the behavior of rust can be changed and/or somebody will want to use a language other than rust. 

Of course, we can impose constraints on wasm modules to be compatible with substrate, however, I think this might be not the decision and could backfire even with Rust.

File: core/executor/src/wasm_executor.rs (hunk 1 above)
Comment (gnunicorn) [reply core/executor/src/wasm_executor.rs]: @pepyakin maybe you could put up a ticket for it that we can link here from? I feel the change is outside of the scope of this PR but the criticism correct and should be resolved (IMHO with an `Err`)

File: core/executor/src/wasm_executor.rs (hunk 6 above)
Comment (arkpar) [reply core/executor/src/wasm_executor.rs]: > Assuming that `used_size` returns the highest written address so far, this will be set to the address of last written byte (shouldn't it be +1?) and then in `Heap` we assume that we can allocate memory starting from this address.

+1 is not required. The buffer should be schrinked to exactly this size.

> 
> This completely ignores the fact that compiler might reserve some space without actually writing to it (aka BSS).
> 

Not exactly. This relies on wasmi to always initialize all data segments, as it does now. Since this code already relies on some of the wasmi internals I don't think this is much of a stretch. 

The bottom line is simple. If the interpreter can do 200bps as demonstrated here it should provide a convenient interface to be used efficiently. I'm sure compatibility can be achieved without sacrificing performance.

Comment (gavofyork) [reply core/executor/src/wasm_executor.rs]: re: the +1 remark, wasmi documentation doesn't appear to be consistent:

```
   /// Returns current used memory size in bytes.
    /// This is the highest memory address that had been written to.
    pub fn used_size(&self) -> Bytes {
        Bytes(self.buffer.borrow().len())
    }
```

i would presume that the `buffer.len()` actually does give a *size*, rather than an *index* (i.e. you write to *index* 0 implies your buffer is *size* 1).

so the remark "This is the highest memory address that had been written to" is wrong: it should read "This is **one more than** the highest memory address that had been written to".

@pepyakin could you confirm?

Comment (arkpar) [reply core/executor/src/wasm_executor.rs]: That function was added by me and the documentation is indeed incorrect.

File: core/executor/src/wasm_executor.rs (hunk 1 above)
Comment (arkpar) [reply core/executor/src/wasm_executor.rs]: Returns error now
